controller/engine_image_controller.go

```diff
@@ -393,73 +393,70 @@ func (ic *EngineImageController) syncNodeDeploymentMap(engineImage *longhorn.Eng
 }
 
 // handleAutoUpgradeEngineImageToDefaultEngineImage automatically upgrades volume's engine image to default engine image when it is applicable
-func (ic *EngineImageController) handleAutoUpgradeEngineImageToDefaultEngineImage() error {
-	concurrentAutomaticEngineUpgradePerNodeLimit, err := ic.ds.GetSettingAsInt(types.SettingNameConcurrentAutomaticEngineUpgradePerNodeLimit)
+func (ic *EngineImageController) handleAutoUpgradeEngineImageToDefaultEngineImage(currentProcessingImage string) error {
+	defaultEngineImage, err := ic.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
 	if err != nil {
 		return err
 	}
-	if concurrentAutomaticEngineUpgradePerNodeLimit <= 0 {
+
+	// To avoid multiple managers doing upgrade at the same time, only allow the
+	// manager that is responsible for the default engine image to do the upgrade
+	if currentProcessingImage != defaultEngineImage {
 		return nil
 	}
 
-	defaultEngineImage, err := ic.ds.GetSettingValueExisted(types.SettingNameDefaultEngineImage)
+	defaultEngineImageResource, err := ic.ds.GetEngineImage(types.GetEngineImageChecksumName(defaultEngineImage))
 	if err != nil {
 		return err
 	}
-	defaultEngineImageResource, err := ic.ds.GetEngineImage(types.GetEngineImageChecksumName(defaultEngineImage))
+
+	concurrentAutomaticEngineUpgradePerNodeLimit, err := ic.ds.GetSettingAsInt(types.SettingNameConcurrentAutomaticEngineUpgradePerNodeLimit)
 	if err != nil {
 		return err
 	}
+	if concurrentAutomaticEngineUpgradePerNodeLimit <= 0 {
+		return nil
+	}
 
 	// List all volumes and select a set of volume for upgrading.
 	volumes, err := ic.ds.ListVolumes()
 	if err != nil {
 		return err
 	}
-	upgradingCandidates, err := ic.getVolumesForEngineImageUpgrading(volumes, defaultEngineImageResource)
-	if err != nil {
-		return err
-	}
 
-	// Make sure not to upgrade too many volumes on the same node at the same time.
-	upgradingCandidates = limitAutomaticEngineUpgradePerNode(volumes, upgradingCandidates, concurrentAutomaticEngineUpgradePerNodeLimit)
+	candidates, inProgress := ic.getVolumesForEngineImageUpgrading(volumes, defaultEngineImageResource)
 
-	// Upgrade engine image for volume
-	for _, v := range upgradingCandidates {
-		oldEngineImage := v.Spec.EngineImage
-		v.Spec.EngineImage = defaultEngineImage
-		v, err = ic.ds.UpdateVolume(v)
-		if err != nil {
-			return err
+	limitedCandidates := limitAutomaticEngineUpgradePerNode(candidates, inProgress, int(concurrentAutomaticEngineUpgradePerNodeLimit))
+
+	for _, vs := range limitedCandidates {
+		for _, v := range vs {
+			ic.logger.WithField("volume", v.Name).WithField("engineImage", v.Spec.EngineImage).Infof("automatically upgrade volume engine image to the default engine image %v", defaultEngineImage)
```

Comment:
You can directly use `ic.logger.WithFields(logrus.Fields{"volume": v.Name, "engineImage": v.Spec.EngineImage})`

Reply:
Thank you! Fixed